networkx/algorithms/tree/mst.py

```diff
@@ -905,6 +905,8 @@ def spanning_tree_total_weight(G, weight):
             #    multiplicative total spanning tree weight if the weight of each edge
             #    is assumed to be 1, which is conveniently built into networkx already,
             #    by calling total_spanning_tree_weight with weight=None
+            #
+            # 3. There are no edges in the graph. Then the weight of the MST is zero
```

Comment:
A MST might have edges and also have weight=0(bcoz of negative weights). So, it might be more appropriate to raise an exception here. Let me know your thoughts.

Reply:
Yes an exception may be useful especially for disconnected graphs.
But I think random_spanning_tree() assumes a connected graph as input. The main loop of random_spanning_tree() has a break condition of len(U) == G.number_of_nodes()-1, which is a tree only if G is connected in the first place. Do we stick to this assumption for now? If so an exception may not be needed (unless there are other failure modes, like the other parts of the loop giving a disconnected graph later, even if we gave a connected graph)

Reply:
I think the disconnected graphs are being handled. And, I was suggesting that when there are no edges in the graph(except for single-node graphs), then it would be more appropriate to raise an exception. Let me know your thoughts on this.

Reply:
random_spanning_tree() afaik doesn't handle disconnected graphs (it raises an exception at the end of the function). For a disconnected graph, suppose there are k components. Then the size of a spanning forest would be n-k (a spanning tree for each component). But random_spanning_tree() returns from within the main loop only when len(U) == G.number_of_nodes()-1, which is n-1, the spanning tree size in a connected graph. For a disconnected graph, it will raise an exception after the loop ("only len(U) edges"). I guess the original author of random_spanning_tree() intended it to work for connected graphs. 

For spanning_tree_total_weight(), we need to decide whether spanning forests are ok as an input to spanning_tree_total_weight(). If they are ok, then a spanning forest for an edgeless graph would just be each node as a singleton tree and the weight would be zero. If they are not ok, we need to check for connectedness since otherwise a spanning tree won't exist. I was reading total_spanning_tree_weight() -- it computes the total weight using a graph Laplacian determinant det(G_laplacian[1:, 1:]), but that is true only for a connected graph (the 1: is specific to connected graphs). [Wiki -- Proof outline](https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem#Proof_outline)
I couldn't find a source for doing the same with disconnected graphs, but I assume it will need component wise determinants 1: and adding or multiplying them up. 

tl;dr disconnected graphs are not currently handled by random_spanning_tree(). Exception when encountering the edgeless case -- sure, but the larger case of disconnected graphs is also not handled, since total_spanning_tree_weight() 's formula works only for connected graphs. Do we check for connectedness and raise Exception?

Reply:
Raising an exception at the end is checking for a connected graph. The message could be better, but it does the check...  This is written for spanning trees, not spanning forests.

`spanning_tree_total_weight` should return zero when there are no spanning trees -- which this does.

Reply:
Got it. Should I expand on the cases in the comment? As below

"3. There are no edges in the graph. Then this function returns zero. Cases where this happens -- (a) single node graph whose spanning tree is that node, (b) > 2 nodes, no spanning tree exists"

Or as below

"3. There are no edges in the graph. Then this function returns zero"

Reply:
Yeah -- this comments doesn't fit well... The code only considers two cases. Let's keep it that way. The second case could be explained as something like (which also fixes some typos/grammar in case 2):
```
            # 2. There are no edges or two or more edges in the graph. Then, we find the
            #    total weight of the spanning trees using the formula in the
            #    reference paper: take the weight of each edge and multiply it by
            #    the number of spanning trees which include that edge. This
            #    can be accomplished by contracting the edge and finding the
            #    multiplicative total spanning tree weight if the weight of each edge
            #    is assumed to be 1, which is conveniently built into networkx already,
            #    by calling total_spanning_tree_weight with weight=None. Note that
            #    Note that with no edges the returned value is just zero.
```